flink-staging/flink-streaming/flink-streaming-core/src/test/java/org/apache/flink/streaming/api/functions/source/SocketTextStreamFunctionTest.java

```diff
@@ -0,0 +1,176 @@
+ */
+public class SocketTextStreamFunctionTest{
+
+	final AtomicReference<Throwable> error = new AtomicReference<Throwable>();
+	private final String host = "127.0.0.1";
+
+	SourceFunction.SourceContext<String> ctx = new SourceFunction.SourceContext<String>() {
+		public String result;
+
+		@Override
+		public void collect(String element) {
+			result = element;
+		}
+
+		@Override
+		public String toString() {
+			return this.result;
+		}
+
+		@Override
+		public void collectWithTimestamp(String element, long timestamp) {
+
+		}
+
+		@Override
+		public void emitWatermark(Watermark mark) {
+
+		}
+
+		@Override
+		public Object getCheckpointLock() {
+			return null;
+		}
+
+		@Override
+		public void close() {
+
+		}
+	};
+
+	public SocketTextStreamFunctionTest() {
+	}
+
+	class SocketSource extends Thread {
+
+		SocketTextStreamFunction socketSource;
+
+		public SocketSource(ServerSocket serverSo, int maxRetry) {
+			this.socketSource =  new SocketTextStreamFunction(host, serverSo.getLocalPort(), '\n', maxRetry);
+		}
+
+		public void run() {
+			try {
+				this.socketSource.open(new Configuration());
+				this.socketSource.run(ctx);
+			}catch(Exception e){
+				error.set(e);
+			}
+		}
+
+		public void cancel(){
+			this.socketSource.cancel();
+		}
+	}
+
+	@Test
+	public void testSocketSourceRetryForever() throws Exception{
+		ServerSocket serverSo = new ServerSocket(0);
+		SocketSource source = new SocketSource(serverSo, -1);
+		source.start();
+
+		int count = 0;
+		Socket channel;
+		while (count < 100) {
+			channel = serverSo.accept();
+			count++;
+			channel.close();
+		}
+		sleep(10000);
+		source.cancel();
+
+		if (error.get() != null) {
+			Throwable t = error.get();
+			t.printStackTrace();
+			fail("Error in spawned thread: " + t.getMessage());
+		}
+
+		assertEquals(100, count);
+	}
+
+	@Test
+	 public void testSocketSourceRetryTenTimes() throws Exception{
+		ServerSocket serverSo = new ServerSocket(0);
+		SocketSource source = new SocketSource(serverSo, 10);
+
+		assertEquals(0, source.socketSource.retries);
+
+		source.start();
+
+		Socket channel;
+		channel = serverSo.accept();
+		channel.close();
+		serverSo.close();
+		while(source.socketSource.retries < 10){
+			long lastRetry = source.socketSource.retries;
+			sleep(10000);
+			assertTrue(source.socketSource.retries > lastRetry);
+		};
+		assertEquals(10, source.socketSource.retries);
+		source.cancel();
+
+		if (error.get() != null) {
+			Throwable t = error.get();
+			t.printStackTrace();
+			fail("Error in spawned thread: " + t.getMessage());
+		}
+
+		assertEquals(10, source.socketSource.retries);
+	}
+
+	@Test
+	public void testSocketSourceNeverRetry() throws Exception{
+		ServerSocket serverSo = new ServerSocket(0);
+		SocketSource source = new SocketSource(serverSo, 0);
+		source.start();
+
+		Socket channel;
+		channel = serverSo.accept();
+		channel.close();
+		serverSo.close();
+		sleep(10000);
```

Comment:
Did you have to add the sleep here to make the test pass? 10 seconds is quite long.